src/mlpack/core/optimizers/sgd/sgd_impl.hpp

```diff
@@ -109,8 +116,17 @@ double SGD<UpdatePolicyType>::Optimize(
     else
       function.Gradient(iterate, currentFunction, gradient);
 
-    // Time to do something really dirty.
-    gradient.transform( [](double val) { return std::min(std::max(val, -3.), 3.); } );
+    // Clip the gradient.
+    if (clipGradient)
+    {
+      gradient.transform
+      (
+        [&](double val)
+        {
+          return std::min(std::max(val, minGradient), maxGradient);
```

Comment:
Honestly I think a better design might be to either contain this logic in the `UpdatePolicy` itself, or make an `UpdatePolicy` wrapper for clipping:

```
template<typename UpdatePolicy>
class GradientClipping
{
 public:
  GradientClipping(const double minGradient, const double maxGradient, UpdatePolicy updatePolicy);

  ...;
};
```

Then you just wrap the desired `UpdatePolicy`:

```
SGD<GradientClipping<UpdatePolicy>> sgd(..., GradientClipping<UpdatePolicy>(2.0, -2.0, UpdatePolicy(...)));
```

What do you think?  This allows us to avoid the gradient checks entirely when the user doesn't want clipping, at compile time.

Also, I agree, I think this should be in a separate PR and we can discuss there.  We should try and limit the scope of this one so that we can merge it soon. :)